db/db_impl_compaction_flush.cc

```diff
@@ -138,7 +138,7 @@ Status DBImpl::FlushMemTableToOutputFile(
 
   Status s;
   if (logfile_number_ > 0 &&
-      versions_->GetColumnFamilySet()->NumberOfColumnFamilies() > 0) {
```

Comment:
@siying and I discussed this a year or so ago. We thought it's a typo but weren't sure if it's safe to change. The case I'm worried about is having a hole after crash-recovery. In particular imagine:

- WAL1 begins flush
- We switch to WAL2 and don't fsync WAL1
- System crashes before finishing flush
- System restarts and reads WAL1 and WAL2. WAL1 may be truncated and WAL2 may be nonempty. Could it introduce a hole in the data post-recovery? I am not sure, tbh, but haven't had time to prove it either way.

Reply:
But if you do prove it, this can significantly improve write perf for single-CF, buffered I/O use case. Then the WAL is essentially free as it can live and die almost entirely in page cache.

Reply:
I also tried to fix this last year but got stuck with fault injection test failures that you are also seeing here. There is some test utility functions that you may also want to fix